### PR TITLE
Hyperball optimizers

### PR DESCRIPTION
**Pull request by WhenWen**
This pull request introduces two new optimizer variants, AdamH and MuonH, and integrates them into the optimizer configuration system. These optimizers are designed to keep the norm of linear weights constant during training, using specialized update rules. Additionally, a new option for QK normalization is added to the Llama model configuration. Below are the most important changes:

**New Optimizer Implementations and Integration:**

* Added `AdamHConfig` and its implementation in `src/levanter/optim/adamh.py`, providing an Adam variant that maintains constant norm for linear weights. Also includes logic for selectively applying AdamH or Adam to parameters based on type and path.
* Added `MuonHConfig` and its implementation in `src/levanter/optim/muonh.py`, introducing a Muon optimizer variant with norm-preserving updates.
* Registered `AdamHConfig` and `MuonHConfig` in the optimizer configuration system and updated `src/levanter/optim/__init__.py` to expose these optimizers. [[1]](diffhunk://#diff-618d7cfc86ebc9661ee125248a6c698d8e166425444a7f6b7e3fb1d1849004bbR15) [[2]](diffhunk://#diff-618d7cfc86ebc9661ee125248a6c698d8e166425444a7f6b7e3fb1d1849004bbR25) [[3]](diffhunk://#diff-618d7cfc86ebc9661ee125248a6c698d8e166425444a7f6b7e3fb1d1849004bbR50-R54)

The new optimizer rules are as following, 

Define normalization of a matrix under the Frobenius norm as $\mathrm{Normalize}(X) = X / \|X\|_F$ and an inner optimizer (such as Adam or Muon) provides update $U_t$, then the next update will be:

$$ W_{t + 1} = \mathrm{Normalize}(W_t - \eta_t  \mathrm{Normalize}(U_t) \|W_0\|_F) \|W_0\|_F. $$

**Model Configuration Enhancements:**

* Added `use_qk_norm` boolean option to `LlamaConfig` in `src/levanter/models/llama.py`, enabling QK normalization in the attention mechanism. This is passed to the attention configuration if enabled. [[1]](diffhunk://#diff-a0809e049aa0f4fda93a804861bd017ace0dbcec88eff2ff5983a7d87db0233fR69) [[2]](diffhunk://#diff-a0809e049aa0f4fda93a804861bd017ace0dbcec88eff2ff5983a7d87db0233fR237)

### CODE REVIEW

File: src/levanter/optim/adamh.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                # this is the case for no layer stacking
+                new_p = p - learning_rate * u * jnp.linalg.norm(p) / jnp.linalg.norm(u)
+                return new_p / jnp.linalg.norm(new_p) * jnp.linalg.norm(p) - p
+            else:
+                axes = tuple(range(1, p.ndim))
+                p_norm = jnp.sqrt(jnp.sum(jnp.square(p), axis=axes, keepdims=True))
+                u_norm = jnp.sqrt(jnp.sum(jnp.square(u), axis=axes, keepdims=True))
+                new_p = p - learning_rate * u * p_norm / u_norm
+                new_p_norm = jnp.sqrt(jnp.sum(jnp.square(new_p), axis=axes, keepdims=True))
+                return new_p / new_p_norm * p_norm - p
```
[the remark this answers is not in the document]
Comment (WhenWen) [reply src/levanter/optim/adamh.py]: This can be ignored as the weight norm are always far from zero.

File: src/levanter/optim/muonh.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            muon_updates,
+            is_leaf=lambda x: x is None,
+        )
+
+        return muonh_updates, ScaleByMuonHState(momentum_buffer=buf)
+
+    return optax.GradientTransformation(init_fn, update_fn)
+
+
+def zeropower_via_newtonschulz5(X, steps=5, eps=1e-7):
```
Comment (dlwh) [inline src/levanter/optim/muonh.py]: can we import the plain muon one?

File: src/levanter/optim/muonh.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                # this is the case for no layer stacking
+                new_p = p - learning_rate * u * jnp.linalg.norm(p) / jnp.linalg.norm(u)
+                return new_p / jnp.linalg.norm(new_p) * jnp.linalg.norm(p) - p
+            else:
+                axes = tuple(range(1, p.ndim))
+                p_norm = jnp.sqrt(jnp.sum(jnp.square(p), axis=axes, keepdims=True))
+                u_norm = jnp.sqrt(jnp.sum(jnp.square(u), axis=axes, keepdims=True))
+                new_p = p - learning_rate * u * p_norm / u_norm
+                new_p_norm = jnp.sqrt(jnp.sum(jnp.square(new_p), axis=axes, keepdims=True))
+                return new_p / new_p_norm * p_norm - p
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/optim/muonh.py]: seems plausible to me? wdyt?

File: src/levanter/optim/adamh.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/optim/adamh.py]: can we clip or you think nah?

File: src/levanter/models/llama.py [hunk 4]
```diff
@@ -66,6 +66,7 @@ class LlamaConfig(HFCompatConfig):
     layer_norm_epsilon: float = 1e-5
     tie_word_embeddings: bool = False
     hybrid_norm: bool = False
+    use_qk_norm: bool = False
```
Comment (rjpower) [inline src/levanter/models/llama.py]: we should have a comment for this, and it looks like we're missing a conversion in the to_hf_config?

File: src/levanter/optim/adamh.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+
+        # projected training for linear weight
+        def scale_invariant_update(p, u):
+            if p is None:
+                return None
+            if p.ndim == 2:
+                # this is the case for no layer stacking
+                new_p = p - learning_rate * u * jnp.linalg.norm(p) / jnp.linalg.norm(u)
+                return new_p / jnp.linalg.norm(new_p) * jnp.linalg.norm(p) - p
```
[the remark this answers is not in the document]
Comment (rjpower) [reply src/levanter/optim/adamh.py:159]: Consider just doing something like max(1e-6, ...) to silence this (or add epsilon, your choice)

If the norm happens to drift low (I don't know, maybe we're working in fp8 or something), it sucks to debug NaNs

File: src/levanter/models/llama.py (hunk 4 above)
Comment (rjpower) [reply src/levanter/models/llama.py]: I don't necessarily like hybrid_norm and use_qk_norm as booleans, but I guess they can be applied independently. In general I like to see things restructured into an enum or configuration:

```
class NormType(Enum):
   DEFAULT_NORM
   QK_NORM
   WHATEVER_NEW_NORM
```

(Probably not for this PR, just an observation).

File: src/levanter/optim/adamh.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (WhenWen) [reply src/levanter/optim/adamh.py:159]: Got it, done now

File: src/levanter/models/llama.py [hunk 6]
```diff
@@ -65,7 +65,12 @@ class LlamaConfig(HFCompatConfig):
     initializer_range: float = 0.02
     layer_norm_epsilon: float = 1e-5
     tie_word_embeddings: bool = False
-    hybrid_norm: bool = False
+    hybrid_norm: bool = (
+        False  # if set to True, this will add an additional layer norm after the attention and MLP layers
+    )
+    use_qk_norm: bool = (
+        False  # if set to True, this will add an additional layer norm after the key and query projections
+    )
```
Comment (dlwh) [inline src/levanter/models/llama.py:73]: we should migrate this work to hackable at some point

File: src/levanter/optim/adamh.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+@OptimizerConfig.register_subclass("adamH")
+@dataclass(frozen=True)
+class AdamHConfig(OptimizerConfig):
+    """
+    This is a variant of the Adam optimizer configuration.
+
+    We ensure that the linear weights stay exactly constant norm as initialization by applying the following update rule:
+
+    p_new_intermediate = p - learning_rate * u * norm(p) / norm(u)
+    p_new = p_new_intermediate / norm(p_new_intermediate) * norm(p)
```
Comment (dlwh) [inline src/levanter/optim/adamh.py:29]: big fan of this idea! (I still want to try (softly) moment matching the initial gaussian)

File: src/levanter/optim/adamh.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        return optax.inject_hyperparams(optimizer)(learning_rate=learning_rate_schedule, adam_lr=adam_lr_schedule)
+
+    def create_mask(self, params):
+        """
+        Creates a mask that labels parameters as 'adamh' or 'adam' based on their
+        dimensionality and module path, using Adam for LayerNorm Gamma and Embedding, and AdamH for all Linear parameters.
+        """
+        paths = leaf_key_paths(params)
+
+        def mask_fn(param, path):
```
Comment (dlwh) [inline src/levanter/optim/adamh.py:86]: ideally we could configure this but let's make sure it works first I guess

**Review by dlwh — APPROVED** [review]
